### ISSUE TITLE
segfault on icalcomponent_vanew with lastmodified property

### ISSUE THREAD

**Issue by sdalu** [libical/libical#585]
System is running FreeBSD 13.1
Call to `icalcomponent_vanew` segfault (SIGSEGV) when lastmodified property is present

Problem was initially reported to cyrus imap. 
For full details see: https://github.com/cyrusimap/cyrus-imapd/issues/4216#issuecomment-1227284115

**Comment by dilyanpalauzov** [comment]
From the above link, running `i.c`:
```c
#include <libical/ical.h>

int main() {
    icalcomponent_vanew(ICAL_VCALENDAR_COMPONENT,
      icalproperty_new_version("2.0"),
      icalproperty_new_prodid("PROD-ABC"),
      icalproperty_new_uid("1234abcd"),
      icalproperty_new_lastmodified(
        icaltime_from_timet_with_zone(1661280150, 0, NULL)),
      icalproperty_new_name("name1"),
      0);
    return 0;
}
```
with
> cc $(pkg-config --cflags libical --libs libical) -o libical-test i.c
> valgrind --leak-check=full --num-callers=40 --track-origins=yes --read-var-info=yes --show-leak-kinds=all --track-fds=yes --error-limit=no --trace-children=yes ./libical-test

prints:
```
==13803== Memcheck, a memory error detector
==13803== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==13803== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==13803== Command: ./libical-test
==13803== Parent PID: 95778
==13803== 
==13803== Invalid read of size 1
==13803==    at 0x48536D4: strcmp (in /usr/local/libexec/valgrind/vgpreload_memcheck-amd64-freebsd.so)
==13803==    by 0x48A55FE: icalcomponent_vanew (in /usr/local/lib/libical.so.3.0.8)
==13803==    by 0x201BA8: main (in /root/libical-test)
==13803==  Address 0x800000000 is not stack'd, malloc'd or (recently) free'd
==13803== 
==13803== 
==13803== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==13803==  Access not within mapped region at address 0x800000000
==13803==    at 0x48536D4: strcmp (in /usr/local/libexec/valgrind/vgpreload_memcheck-amd64-freebsd.so)
==13803==    by 0x48A55FE: icalcomponent_vanew (in /usr/local/lib/libical.so.3.0.8)
==13803==    by 0x201BA8: main (in /root/libical-test)
==13803==  If you believe this happened as a result of a stack
==13803==  overflow in your program's main thread (unlikely but
==13803==  possible), you can try to increase the size of the
==13803==  main thread stack using the --main-stacksize= flag.
==13803==  The main thread stack size used in this run was 16777216.
==13803== 
```

**Comment by winterz** [comment]
I can reproduce on linux (Fedora) with ASAN builds

**Comment by winterz** [comment]
whereas this code seems to work fine.  so I suspect a subtle varargs problem.  could you try this code snippet, in the meantime?

```
    struct icaltimetype lm = icaltime_from_timet_with_zone(1661280150, 0, NULL);
    icalcomponent *comp = icalcomponent_new(ICAL_VCALENDAR_COMPONENT);
    icalcomponent_add_property(comp, icalproperty_new_version("2.0"));
    icalcomponent_add_property(comp, icalproperty_new_prodid("PROD-ABC"));
    icalcomponent_add_property(comp, icalproperty_new_uid("1234abcd"));
    icalcomponent_add_property(comp, icalproperty_new_lastmodified(lm));
    icalcomponent_add_property(comp, icalproperty_new_name("name1"));
```

**Comment by winterz** [comment]
FTR: reproduced on Mac with Apple Clang.

**Comment by dilyanpalauzov** [comment]
Ideally add `-g` to `cc`:

> cc -g $(pkg-config --cflags libical --libs libical) -o libical-test i.c

and install debug data for libical, so that valgrind can produce more useful data.

**Comment by winterz** [comment]
oh my. you need to pass "NULL" as the final argument, not 0.

ie.
```
#include <libical/ical.h>

int main() {
    icalcomponent_vanew(ICAL_VCALENDAR_COMPONENT,
      icalproperty_new_version("2.0"),
      icalproperty_new_prodid("PROD-ABC"),
      icalproperty_new_uid("1234abcd"),
      icalproperty_new_lastmodified(
        icaltime_from_timet_with_zone(1661280150, 0, NULL)),
      icalproperty_new_name("name1"),
      NULL);
    return 0;
}
```

If anyone has ideas on how to force C to barf if NULL isn't passed, I'm all ears.
my C is quite rusty by now.

for now I can add documentation.

**Comment by dilyanpalauzov** [comment]
How is `NULL` different from `0`?

**Comment by sdalu** [comment]
@dilyanpalauzov `0` is an `int`, `NULL` is a `void *`. On modern computer `sizeof(int)` is 4, `sizeof(void *)` is 8. As it is a variadic function there is no type promotion for element of this size.
@winterz I thinks there is no way, to have the compiler perform type checking on variadic function except for the one defined in `__attribute__((format (...)))` which are related to the `printf` family.

**Comment by dilyanpalauzov** [comment]
Perhaps by changing src/libical/icalcomponent.h from
> LIBICAL_ICAL_EXPORT icalcomponent *icalcomponent_vanew(icalcomponent_kind kind, ...);

to
> LIBICAL_ICAL_EXPORT icalcomponent *icalcomponent_vanew(icalcomponent_kind kind, ...) \_\_attribute__((sentinel));

as suggested at https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-sentinel-function-attribute . I have not tried to see how this attribute works in practice.

**Comment by dilyanpalauzov** [comment]
The example in examples/access_components.c:211 does finish  icalproperty_vanew_organizer(…), icalproperty_vanew_attendee(…), icalproperty_vanew_dtstart(…) with 0, not with NULL.

**Comment by winterz** [comment]
> The example in examples/access_components.c:211 does finish icalproperty_vanew_organizer(…), icalproperty_vanew_attendee(…), icalproperty_vanew_dtstart(…) with 0, not with NULL.

indeed. I'll change those to NULL

**Comment by dilyanpalauzov** [comment]
doc/UsingLibical.md says:
```
If any of the constructors fail, they will return 0. 
```
Should this be changed to return NULL?

Moreover src/libical/icalcomponent.c contains:
```
icalcomponent *icalcomponent_vanew(icalcomponent_kind kind, ...) {
  icalcomponent *impl = icalcomponent_new_impl(kind);
  if (impl == 0) return 0;
```
Provided that NULL is a pointer, an 0 is not a pointer, but is casted to `(icalcomponent *)0`, the above code might be made more consequest by changing it to `if (impl == NULL) return NULL;`.

**Comment by dilyanpalauzov** [comment]
Executing `git grep -nA20 _vanew` shows more places still containing …_vanew(…, 0): examples/access_components.c lines 248 and 250; examples/access_properties_and_parameters.c-120 as (void*)0, src/test/stow.c lines 304, 305; in src/test/regression.c and so on.

**Comment by winterz** [comment]
@dilyanpalauzov  mind taking a review of https://github.com/libical/libical/pull/661 (for the 3.0 branch)

I'll merge into master once this lands and fix any other instances there

**Comment by dilyanpalauzov** [comment]
The single _vanew call I found with terminating `0`, not casted to `void*`, on https://github.com/libical/libical/tree/work/winterz/voidstar is in src/test/regression-component.c:340-341:
```c
icalproperty_vanew_dtend(icaltime_from_timet_with_zone(tm2, 0, azone),
                         icalparameter_new_tzid("America/Los_Angeles"), 0),

```

**Comment by winterz** [comment]
thank you for reviewing. that fix should be in src/test/regression-component.c now
